File: merge.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void intercala(int inicio, int meio, int fim, int v[]) {
    int i = inicio, j = meio + 1, k = 0;
    int *aux = malloc((fim - inicio + 1) * sizeof(int));

    while (i <= meio && j <= fim) {
        if (v[i] <= v[j]) aux[k++] = v[i++];
        else              aux[k++] = v[j++];
    }
    while (i <= meio) aux[k++] = v[i++];
    while (j <= fim)  aux[k++] = v[j++];
    for (i = inicio, k = 0; i <= fim; i++, k++)
        v[i] = aux[k];

    free(aux);
}

void mergesort(int inicio, int fim, int v[]) {
    if (inicio < fim) {
        int meio = (inicio + fim) / 2;
        mergesort(inicio, meio, v);
        mergesort(meio + 1, fim, v);
        intercala(inicio, meio, fim, v);
    }
}

int main() {
    FILE *f = fopen("../casos_de_teste/melhor-caso-500k.txt", "r"); //mude o arquivo a ser ordenado aqui
    if (!f) {
        perror("Erro ao abrir arquivo");
        return EXIT_FAILURE;
    }

    const int capacidade = 1000001;
    int *v = malloc(capacidade * sizeof(int));
    if (!v) {
        perror("Erro de alocação");
        fclose(f);
        return EXIT_FAILURE;
    }

    int tamanho = 0;
    while (tamanho < capacidade && fscanf(f, "%d", &v[tamanho]) == 1)
        tamanho++;
    fclose(f);

    if (tamanho == 0) {
        fprintf(stderr, "Nenhum número lido do arquivo ou formato inválido\n");
        free(v);
        return EXIT_FAILURE;
    }

    clock_t inicio = clock();

    mergesort(0, tamanho - 1, v);

    clock_t fim = clock();

    double duracao = (double)(fim - inicio) / CLOCKS_PER_SEC;
    printf("Tempo de ordenação: %.6f segundos\n", duracao);

    //printf("Vetor ordenado com %d números:\n", tamanho);
    //for (int i = 0; i < tamanho; i++)
    //    printf("%d ", v[i]);
    //printf("\n");

    free(v);
    return EXIT_SUCCESS;
}

File: merge.py
import time
import sys

def intercala(v, inicio, meio, fim):
    """Intercala as sublistas v[inicio:meio+1] e v[meio+1:fim+1]."""
    esquerda = v[inicio:meio+1]
    direita = v[meio+1:fim+1]
    i = j = 0
    k = inicio

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] <= direita[j]:
            v[k] = esquerda[i]
            i += 1
        else:
            v[k] = direita[j]
            j += 1
        k += 1

    while i < len(esquerda):
        v[k] = esquerda[i]
        i += 1
        k += 1
    while j < len(direita):
        v[k] = direita[j]
        j += 1
        k += 1

def mergesort(v, inicio, fim):
    """Ordena v[inicio:fim+1] recursivamente."""
    if inicio < fim:
        meio = (inicio + fim) // 2
        mergesort(v, inicio, meio)
        mergesort(v, meio + 1, fim)
        intercala(v, inicio, meio, fim)

def main():
    filename = "../trabalho-teoria/casos_de_teste/melhor-caso-100k.txt" 
    capacidade = 1_000_000

    try:
        with open(filename, "r") as f:
            v = []
            for linha in f:
                if len(v) >= capacidade:
                    break
                for token in linha.split():
                    if len(v) >= capacidade:
                        break
                    try:
                        v.append(int(token))
                    except ValueError:
                        print(f"Aviso: formato inválido em '{token}', pulando.", file=sys.stderr)
    except OSError as e:
        print(f"Erro ao abrir arquivo: {e}", file=sys.stderr)
        sys.exit(1)

    if not v:
        print("Nenhum número lido do arquivo ou formato inválido.", file=sys.stderr)
        sys.exit(1)

    # Ordenação e medição de tempo
    inicio = time.perf_counter()
    mergesort(v, 0, len(v) - 1)
    fim = time.perf_counter()

    duracao = fim - inicio
    print(f"Tempo de ordenação: {duracao:.6f} segundos")

    # Se quiser ver o vetor ordenado, descomente:
    # print(f"Vetor ordenado com {len(v)} números:")
    # print(" ".join(map(str, v)))

if __name__ == "__main__":
    main()
